serde: support std::chrono::duration

bharathv opened the pull request:
## Cover letter

Adds support for duration and time_point.

Notes:
* `std::chrono::duration` is serialized with a nanosecond precision (int64_t) and cast-ed back to the type provided during deserialiation. This makes sure that we have a consistent on disk representation irrespective of subtle type re-definitions in the layers above. Audited all serde callers in `21.11.x` and `22.1.x` for compat issues with existing timestamp serialization and found that two use cases (`cloudstorage::partition_manifest::segment_meta` and `storage::index_state`) do it and both of them use `model::timestamp` which treats timestamps as int64_t, so no backward compatibility issues here.
* Added a concept check to discourage future developers from serializing non-system clock time points.

## Release notes

* none

bharathv commented:
This one seems like a legit failure with overflows, putting together a fix. chrono types has some weird behavior w.r.t overflows.

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -603,6 +613,10 @@ void read_nested(iobuf_parser& in, T& t, std::size_t const bytes_left_limit) {
     } else if constexpr (is_duration_v<Type>) {
         auto rep = read_nested<int64_t>(in, bytes_left_limit);
         t = std::chrono::duration_cast<Type>(std::chrono::nanoseconds{rep});
+    } else if constexpr (is_time_point_v<Type>) {
+        auto duration = read_nested<typename Type::duration>(
+          in, bytes_left_limit);
+        t = std::chrono::time_point<typename Type::clock>(std::move(duration));
```
like std::pair, should we support time point as a native type? imo it is a red flag when i see time points being sent over the wire or stored on disk. i suppose there are legitimate use cases, but maybe those should either be handled explicitly by the caller or we could create a serde::time_point type that forces users to think about what they are doing.

bharathv replied:
You mean it is not logical (no pun intended) due to clock sync issues? (like the general notion of clock drifts)

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -279,8 +286,6 @@ void write(iobuf& out, T t) {
         return write(out, static_cast<typename Type::type>(t));
     } else if constexpr (reflection::is_ss_bool_v<Type>) {
         write(out, static_cast<int8_t>(bool(t)));
-    } else if constexpr (std::is_same_v<Type, std::chrono::milliseconds>) {
```
>In https://github.com/redpanda-data/redpanda/pull/5427 @BenPope rightly pointed that this is not backward compatible since we change the on disk/wire serialization but @dotnwat mentioned that no one may be using this (yet) given the framework is fairly new (need to verify).

correct. this needs to be researched and verified, but does seem like a viable path forward.

this needs to be done in a the next week or two before the release.

bharathv replied:
I've gone thru all serde callers in `21.11.x` and `22.1.x`, only interesting I could find are `cloudstorage::partition_manifest::segment_meta` and  `storage::index_state`  and both of them use `model::timestamp `which treats timestamps as int64_t, so I don't think we are breaking any compat here.

graphcareful reviewed (commented):
LGTM, I however am questioning the use of sending a time_point over the wire. If its not needed IMO we shouldn't have support for it

bharathv commented:
> I however am questioning the use of sending a time_point over the wire. If its not needed IMO we shouldn't have support for it

I agree about the implications of serializing time points and hence the wrapper around it, so those using it are likely aware of the downsides. This is more of a convenience thing though, For the context here, I did it to serialize some state in [rm_stm](https://github.com/redpanda-data/redpanda/blob/de6bcdb6e196e274ccc7532e6658599739e2daab/src/v/cluster/rm_stm.h#L200)  without much effort.

graphcareful reviewed (approved):
LGTM, pending maybe another look by @BenPope

BenPope commented on src/v/serde/serde.h:
```diff
@@ -403,6 +411,8 @@ void write(iobuf& out, T t) {
         auto d = std::chrono::duration_cast<std::chrono::nanoseconds>(t);
         write<int64_t>(out, d.count());
+    } else if constexpr (is_time_point_v<Type>) {
+        write(out, t.time_since_epoch());
```
This doesn't seem to cast the underlying type. Is this in `std::nanoseconds` or `duration::rep`?

bharathv replied:
It doesn't matter, right? (unless I'm misunderstanding your question). Its like serializing any duration, the cast to ns happens in the is_duration_v<> branch.

BenPope replied:
`t.time_since_epoch()` will serialise in units of `std::duration_cast<T::rep>`. For `std::chrono::duration` we chose `std::chrono::nanonseconds`. So I wonder why the disparity.

BenPope replied:
Oh, I see, it's forwarded to the `std::duration` overload?

bharathv replied:
Right. Eventually it is casted to std::chrono::nanoseconds.

BenPope commented on src/v/serde/test/serde_test.cc:
```diff
@@ -853,11 +854,12 @@ SEASTAR_THREAD_TEST_CASE(duration_type_test) {
-    auto yest = now - 24h;
-    auto now_steady = std::chrono::steady_clock::now();
 
+    serde::time_point now = ss::lowres_clock::now();
```
I'd rather this had an explicit `::rep`

bharathv replied:
Added one more test with a custom rep.

bharathv commented on src/v/serde/serde.h:
```diff
@@ -157,6 +157,13 @@ struct is_absl_node_hash_map<absl::node_hash_map<Args...>> : std::true_type {};
 template<typename T>
 inline constexpr bool is_absl_node_hash_map_v = is_absl_node_hash_map<T>::value;
 
+template<typename T>
+struct is_duration : std::false_type {};
+template<typename... Args>
+struct is_duration<std::chrono::duration<Args...>> : std::true_type {};
+template<typename T>
+inline constexpr bool is_duration_v = is_duration<T>::value;
```
I copy pasted for consistency. Is it ok to define just this one with concepts? (double checking before I redo this).

BenPope replied:
I'm like 2/10 on this. If it works, it works. This is the context: https://github.com/redpanda-data/redpanda/pull/5288

bharathv replied:
Can't we just use the builtin std::same_as? This seems to work. LMK if you think otherwise.

```
template<class T>
concept is_duration_v = std::same_as<T, std::chrono::duration<typename T::rep, typename T::period>>;
```

BenPope replied:
Yeah, that works. The `_v` is unusual for a concept, and I think `chrono` should be in the name.

bharathv commented on src/v/serde/serde.h:
```diff
@@ -168,10 +175,10 @@ inline constexpr auto const is_serde_compatible_v
     || std::is_same_v<T, iobuf>
     || std::is_same_v<T, ss::sstring>
     || std::is_same_v<T, bytes>
+    || is_duration_v<T>
```
Order not important, I just grouped all the "is_" together, perhaps I should move it a couple of places below to maintain sortedness.

BenPope replied:
No big deal. I just wondered if it was relevant.

bharathv replied:
Fixed the sorted order.

bharathv commented on src/v/serde/serde.h:
```diff
@@ -360,6 +365,44 @@ void write(iobuf& out, T t) {
+        // On a typical platform which uses int64_t for 'rep', it roughly
+        // translates to ~292 years.
+        using wide_nano_s
+          = std::chrono::duration<double, std::chrono::nanoseconds::period>;
```
Think you answered this below, its not likely we are getting anywhere close to this. Just for overflow check, needed a wider type with more bits.

bharathv commented:
Rebased on top of all the latest serde changes and addressed comments.

bharathv commented on src/v/serde/serde.h:
```diff
@@ -360,6 +365,44 @@ void write(iobuf& out, T t) {
+          std::chrono::nanoseconds::min().count());
+        constexpr auto nano_max = static_cast<double>(
+          std::chrono::nanoseconds::max().count());
+        if (unlikely(count < nano_min || count > nano_max)) {
```
> I guess we can't roundtrip std::chrono::milliseconds::max()

Right we have it in the unit tests.

>  Do we ever have one for something like a sentinel value for "no timeout"?

At least in our use cases where durations are mostly positive, I guess we can use a negative (-1) as no timeout? or define a custom sentinel specific to the duration type in use.

bharathv commented:
Failures known issues #5466 #4772. @BenPope this is ready for review again.

BenPope commented on src/v/serde/time_point.h:
```diff
@@ -0,0 +1,50 @@
+// serializing this special wrapped type making it easy for us to audit the
+// usage.
+// A time_point is serialized as the underlying duration epoch.
+template<class time_point_type = ss::lowres_clock::time_point>
```
I guess I just don't really think this is a safe utility: 

[ss::lowres_clock](https://github.com/redpanda-data/seastar/blob/79a3038e90c45195a0446a067f516f7ff0ada0be/include/seastar/core/lowres_clock.hh#L47-L48
)
> This is a monotonic clock with a granularity of ~task_quota. Time points from this clock do not correspond to system time.



One could argue that it could be made safer by requiring the clock itself to have `to_time_t` / `from_time_t`.

BenPope replied:
Made it safer:
```diff
--- a/src/v/serde/time_point.h
+++ b/src/v/serde/time_point.h
@@ -15,13 +15,20 @@
 
 namespace serde {
 
+template<typename T>
+concept system_clock_duration = requires {
+    // std::is_clock<T> - I don't think libc++ has this
+    T::clock::from_time_t;
+    T::clock::to_time_t;
+};
+
 // A serializable time_point type. Generally time points have a risk of
 // clock synchronization issues when deserialized on node with a clock drift.
 // Rather than letting users serialize arbitrary time_points, we only allow
 // serializing this special wrapped type making it easy for us to audit the
 // usage.
 // A time_point is serialized as the underlying duration epoch.
-template<class time_point_type = ss::lowres_clock::time_point>
+template<
+  system_clock_duration time_point_type = ss::lowres_system_clock::time_point>
 struct time_point : envelope<time_point<time_point_type>, version<0>> {
     using duration = typename time_point_type::duration;
     using clock = typename time_point_type::clock;
```

bharathv replied:
Done, switched to lowres_system_clock as default.

BenPope commented on src/v/serde/time_point.h:
```diff
@@ -0,0 +1,59 @@
+// are system clocks.
+template<class T>
+concept is_system_clock = requires {
+    T::clock::to_time_t;
```
It's unfortunate that libc++ doesn't seem to have [std::is_clock`](https://en.cppreference.com/w/cpp/chrono/is_clock), but those checks (or a standalone concept) could be added, too, for completeness.

BenPope commented on src/v/serde/time_point.h:
```diff
@@ -0,0 +1,59 @@
+
+    time_point() = default;
+
+    time_point(time_point_type t)
```
I think I heard that this implicit conversion is intended, which probably deserves some documentation (and  some more thought on my part). If that is the case, does it make sense to have an `operator time_point_type() const`, too?

I guess some usage examples would help, perhaps serde roundtripping an `envelope` that contains a `time_point_type`?
```c++
struct foo : serde_envelope<foo, version<0>> {
   ss::lowres_system_clock::time_point tp;
};
```

BenPope commented on src/v/serde/time_point.h:
```diff
@@ -0,0 +1,59 @@
+// usage.
+// A time_point is serialized as the underlying duration epoch.
+template<is_system_clock time_point_type = ss::lowres_system_clock::time_point>
+struct time_point : envelope<time_point<time_point_type>, version<0>> {
```
Versioning what is usually an int64_t seems like overkill.

BenPope replied:
Maybe we should document the problem, perhaps something like:
```c++
template<T>
requires !is_system_clock time_point_type<T>
struct time_point : envelope<time_point<time_point_type>, version<0>> {
   static_assert(false, "serializing a time_point that does not correspond to wall-clock time is probably not going to do what you want");
}

dotnwat commented:
what did you find out about using nanosecond duration w.r.t. previous version of redpanda that we'll be upgrading from? it'd be great to go to nanoseconds, but we need to make sure its safe. can you add your findings to the cover letter?

BenPope commented on src/v/serde/test/serde_test.cc:
```diff
@@ -840,4 +841,45 @@ SEASTAR_THREAD_TEST_CASE(duration_type_test) {
+    struct outer2 : serde::envelope<outer, serde::version<0>> {
+        clock::time_point tp;
+
+        auto serde_write(iobuf& out) {
+            serde::write(out, serde::time_point(tp));
+        }
+
+        auto serde_read(iobuf_parser& in, const serde::header& h) {
+            tp = serde::read_nested<serde::time_point<clock::time_point>>(
+              in, h._bytes_left_limit);
+        }
```
I'm not sure I understand why there are implicit conversions if it's going to be used explicitly. If it's going to be used explicitly, I don't know if this provides sufficient value.

Do you have any use-cases in real code for `serde<time_point>`?

BenPope commented on src/v/serde/time_point.h:
```diff
@@ -0,0 +1,99 @@
+// When using the underlying timepoint directly, we can serialize it by
+// constructing an intermediate serde::time_point type.
+
+template<is_system_clock time_point_type = ss::lowres_system_clock::time_point>
```
Maybe we should document the problem, perhaps something like:
```c++
template<T>
requires !is_system_clock time_point_type<T>
struct time_point : envelope<time_point<time_point_type>, version<0>> {
   static_assert(false, "serializing a time_point that does not correspond to wall-clock time is probably not going to do what you want");
}

BenPope commented on src/v/serde/time_point.h:
```diff
@@ -0,0 +1,99 @@
+    }
+};
+
+} // namespace serde
```
nit: newline

bharathv commented:
Had a brief chat with @BenPope offline. We decided to undo the time_point commit for now given the inherent risks of serializing time points and rather add a static assert to avoid people doing it. We decided to keep the duration support as is, with ns precision. I'll make the changes and update the PR shortly.

bharathv commented:
Latest force push rebases on dev and hopefully addresses all the comments.

> what did you find out about using nanosecond duration w.r.t. previous version of redpanda that we'll be upgrading from? it'd be great to go to nanoseconds, but we need to make sure its safe. can you add your findings to the cover letter?

Done.

bharathv commented:
Test failure is #4702.

BenPope commented on src/v/serde/time_point.h:
```diff
@@ -0,0 +1,48 @@
+// discourage future serde developers and users to serialize non system clock
+// time points. For discussion see
+// https://github.com/redpanda-data/redpanda/pull/5432
+template<class T>
+concept is_not_system_clock_time_point
+  = !::detail::
+      is_specialization_of_v<T, std::chrono::time_point> || !is_system_clock<T>;
```
This is a little awkward to read with all the negatives.

BenPope reviewed (commented):
How do you feel about this?
```diff
diff --git a/src/v/serde/serde.h b/src/v/serde/serde.h
index 5a2829570..dd79979c0 100644
--- a/src/v/serde/serde.h
+++ b/src/v/serde/serde.h
@@ -144,7 +144,7 @@ concept is_chrono_duration
 
 template<typename T>
 inline constexpr auto const is_serde_compatible_v
-  = is_not_system_clock_time_point<T> && (is_envelope<T>
+  = (is_envelope<T>
     || (std::is_scalar_v<T>  //
          && (!std::is_same_v<float, T> || std::numeric_limits<float>::is_iec559)
          && (!std::is_same_v<double, T> || std::numeric_limits<double>::is_iec559)
@@ -162,6 +162,7 @@ inline constexpr auto const is_serde_compatible_v
     || is_absl_node_hash_set<T>
     || is_absl_node_hash_map<T>
     || is_chrono_duration<T>
+    || is_chrono_time_point<T>
     || is_std_unordered_map<T>
     || is_fragmented_vector<T> || reflection::is_tristate<T> || std::is_same_v<T, ss::net::inet_address>);
 
@@ -393,6 +394,17 @@ void write(iobuf& out, T t) {
         }
         auto d = std::chrono::duration_cast<std::chrono::nanoseconds>(t);
         write<int64_t>(out, d.count());
+    } else if constexpr (is_chrono_time_point<Type>) {
+        static_assert(
+          is_chrono_sys_time<Type>,
+          "time_point::clock doesn't have a consistent epoch; it should not be "
+          "serialized");
+        write(t.time_since_epoch());
+        // TODO: When std::chrono::clock_cast is implemented, remove the assert
+        // and use this:
+        // auto sys_time =
+        //   std::chrono::clock_cast<std::chrono::system_clock>(t);
+        // write(sys_time.time_since_epoch());
     }
 }
 
diff --git a/src/v/serde/time_point.h b/src/v/serde/time_point.h
index 0ddf63ba8..303953f60 100644
--- a/src/v/serde/time_point.h
+++ b/src/v/serde/time_point.h
@@ -36,13 +36,12 @@ concept is_system_clock = is_clock<typename T::clock> && requires {
     T::clock::from_time_t;
 };
 
-// Only system clock time points are serializable, this check is meant to
-// discourage future serde developers and users to serialize non system clock
-// time points. For discussion see
-// https://github.com/redpanda-data/redpanda/pull/5432
 template<class T>
-concept is_not_system_clock_time_point
-  = !::detail::
-      is_specialization_of_v<T, std::chrono::time_point> || !is_system_clock<T>;
+concept is_chrono_time_point
+  = ::detail::is_specialization_of_v<T, std::chrono::time_point>;
+
+template<typename T>
+concept is_chrono_sys_time
+  = is_chrono_time_point<T> && is_system_clock<typename T::clock>;
 
 } // namespace serde
```
Obviously read would need implementing too.

I'm still not entirely convinced I want to implement it, but I think having the overload is useful to prevent people from implementing it.

Maybe we can not implement it by doing:
```c++
    } else if constexpr (is_chrono_time_point<Type>) {
        static_assert(
          !is_chrono_time_point<Type>,
          "time_point::clock doesn't have a consistent epoch; it should not be "
          "serialized");
    }
```
But there's a strong chance somebody will just attempt to workaround it by taking their `time_point` and call `time_since_epoch()` to erase the clock.

bharathv commented:
> How do you feel about this?
> Obviously read would need implementing too.

This is the reason I didn't do it this way (two extra blocks of code). `is_serde_compatible_v` is called in both read/write paths and all the time_point checks sits in a header that very few people care about :-D . But I agree that the static assert error is way more nicer than some template garbage we in error messages. Let me know if you prefer to do this way strongly, I can switch.

> This is a little awkward to read with all the negatives.

Ya I can get rid of double negatives.

BenPope commented:
> > How do you feel about this?
> > Obviously read would need implementing too.
> 
> This is the reason I didn't do it this way (two extra blocks of code). `is_serde_compatible_v` is called in both read/write paths and all the time_point checks sits in a header that very few people care about :-D . 

...and then #include that header in serde.h anyway.

We also pull in just about every container, so that boat has sailed for now.

>But I agree that the static assert error is way more nicer than some template garbage we in error messages. Let me know if you prefer to do this way strongly, I can switch.
> 
> > This is a little awkward to read with all the negatives.
> 
> Ya I can get rid of double negatives.

I think it's nice to have the explanation in the error that's seen, to avoid workarounds that won't work (e.g.: `ss::lowres_clock::time_since_epoch()`), or a dance with `ss::lowres_system_clock::now()`.

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -166,6 +171,20 @@ inline constexpr auto const are_bytes_and_string_different = !(
 
+template<class R, class P>
+bool check_duration_overflow(const std::chrono::duration<R, P>& duration) {
+    // Here we check the duration overflow by casting it into a wider type
+    // (double).
+    using wide_nano_s
+      = std::chrono::duration<double, std::chrono::nanoseconds::period>;
```
I'm not sure what this means. IIUC it's casting into fractional seconds?

bharathv replied:
This is for checking overflows in the duration (see the comment in L366). Since we are serializing as nanoseconds, the max duration (in nanoseconds) we support is nanoseconds::max(). Now std::chrono is weird in this regard, since nanoseconds and milliseconds are both represented by int64_t, nanoseconds::max() == milliseconds::max() == int64_max. Now for example one tries to cast milliseconds::max() into a nanosecond duration, that overflows silently.  We are avoiding that here and throwing if thats the case.
The way we detect the overflow is by casting it into a type that has more bits than  int64_t (which is double) and see if its > nano::max().

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -344,6 +361,37 @@ void write(iobuf& out, T t) {
+        // nanoseconds::max()'s equivalent on the duration to be serialized.
+        // On a typical platform which uses int64_t for 'rep', it roughly
+        // translates to ~292 years.
+        if (unlikely(check_duration_overflow(t))) {
```
Can we remove the floating point arithmetic here? Can we place restrictions on what serde supports to make this more straight forward? Can we use our own sentinel values for min/max representation? Should all callers have to provide nanoseconds and figure it out for themselves?

bharathv replied:
> Can we remove the floating point arithmetic here

I tried an alternative way of detecting overflows by casting to ns and back and see if we have the same value but that doesn't work.

> Can we use our own sentinel values for min/max representation? Should all callers have to provide nanoseconds and figure it out for themselves?

We can't control the `Rep` of duration to use sentinels? This code works with any custom duration type and forcing the callers to pass nanoseconds means we are just forcing them to do the same check and serialization is not seamless.

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -143,7 +144,7 @@ concept is_chrono_duration
 template<typename T>
 inline constexpr auto const is_serde_compatible_v
-  = is_envelope<T>
+  = is_not_system_clock_time_point<T> && (is_envelope<T>
```
looks like this is not used and this patch can be dropped? or, i mean, we don't support time_point yet so what are we protecting against?

BenPope replied:
> looks like this is not used and this patch can be dropped? or, i mean, we don't support time_point yet so what are we protecting against?

The idea is to prevent somebody from implementing time_point for a clock that isn't a system_clock. For example, `ss::lowres_clock`, because separate instances don't share an epoch.

bharathv replied:
The check is for the future developers ...(this can be refactored to what Ben suggested in his previous comment ...with a static assert).

dotnwat reviewed (commented):
In this commit:

https://github.com/redpanda-data/redpanda/commit/2d266913585a5c4f7c2526ab360b42d7be921681

it looks like @BenPope did a duration cast on lots of serde types, presumably because of serastar changing something. what does that mean for this PR, if anything?

bharathv reviewed (commented):
> presumably because of serastar changing something. what does that mean for this PR, if anything?

Its the same trick I'm using here except that I'm serializing as nanoseconds. This approach saves us from such seastar bugs because we have a consistent on wire/disk representation in ns irrespective of such typedef changes.

bharathv commented:
Switched to using a builtin that works with/without `-O2`.. more details here https://godbolt.org/z/f67nEf1E8

bharathv commented:
Clamped to limits with an error log on overflows / underflows.. the log lines look like this.. lets see how many ductake tests run into this..

```
ERROR 2022-08-04 23:13:33,543 [shard 0] serde - serde.h:220 - Overflow or underflow detected when casting to nanoseconds, clamping to a limit. Input: 9223372036854775807  type: std::chrono::duration<long long, std::ratio<1, 1000>>
ERROR 2022-08-04 23:13:33,543 [shard 0] serde - serde.h:220 - Overflow or underflow detected when casting to nanoseconds, clamping to a limit. Input: 9223372036854775807  type: std::chrono::duration<long, std::ratio<3600, 1>>
ERROR 2022-08-04 23:13:33,543 [shard 0] serde - serde.h:220 - Overflow or underflow detected when casting to nanoseconds, clamping to a limit. Input: -9223372036854775808  type: std::chrono::duration<long long, std::ratio<1, 1000>>
ERROR 2022-08-04 23:13:33,543 [shard 0] serde - serde.h:220 - Overflow or underflow detected when casting to nanoseconds, clamping to a limit. Input: -9223372036854775808  type: std::chrono::duration<long, std::ratio<3600, 1>>
```

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -166,6 +174,50 @@ inline constexpr auto const are_bytes_and_string_different = !(
+          "clamping to a limit. Input: {}  type: {}",
+          duration.count(),
+          type_str<input_type>());
+        return duration.count() < 0 ? std::chrono::nanoseconds::min().count()
+                                    : std::chrono::nanoseconds::max().count();
```
clamping to max on overflow seems to make sense. why do you clamp to min if an overflow occurred and the original value was negative?

bharathv replied:
Setting it to max() can result in exactly opposite comparison results (after deserialization) compared to a negative duration?

For ex: say -10 days overflows hypothetically (just picking some smaller value for explanation), 

if casted to max(), say we get 8 days (clamped equivalent) 
if casted to min(), we  get -8 days..

If the caller does comparisons like < / > on the deserialized value, its more closer to -8 days than +8 days? Positive values can probably even result in incorrect checks

dotnwat commented on src/v/serde/test/serde_test.cc:
```diff
@@ -21,6 +21,7 @@
 #include <seastar/net/inet_address.hh>
 #include <seastar/testing/thread_test_case.hh>
 
+#include <__chrono/duration.h>
```
is the header you intended?

bharathv replied:
IDE magic, removed,

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -75,6 +75,9 @@ struct header {
     checksum_t _checksum;
 };
 
+template<typename T>
+concept always_false = false;
+
```
this is defined in the second patch, but used in the first

bharathv replied:
removed this stuff by moving assert to the top.

dotnwat commented on src/v/serde/serde.h:
```diff
@@ -665,6 +673,11 @@ void read_nested(iobuf_parser& in, T& t, std::size_t const bytes_left_limit) {
           "rounding issues.");
         auto rep = read_nested<int64_t>(in, bytes_left_limit);
         t = std::chrono::duration_cast<Type>(std::chrono::nanoseconds{rep});
+    } else if constexpr (is_chrono_time_point<Type>) {
+        static_assert(
+          always_false<Type>,
+          "Time point serialization is risky and can have unintended "
+          "consequences. Check with RedPanda team before fixing this.");
```
one of these is also in the previous commit. perhaps you intended to move it into this commit?

also, why can we not do this at the top of the function:

```
static_assert(!is_chrono_time_point<Type>, "...");
```

it doesn't look like we need to assert false in the true condition.

dotnwat reviewed (commented):
looks good. just a couple minor things

dlex commented on src/v/serde/serde.h:
```diff
@@ -166,9 +177,64 @@ inline constexpr auto const are_bytes_and_string_different = !(
+    static_assert(
+      !is_chrono_time_point<Type>,
+      "Time point serialization is risky and can have unintended "
+      "consequences. Check with RedPanda team before fixing this.");
```
Capitalization: "Redpanda". I'm sorry for this nitpicking, [this](https://vectorizedio.atlassian.net/wiki/spaces/DOC/pages/92635182/Redpanda+Documentation+Style+Guide#Redpanda-specific-word-list) is still fresh on my memory :)

bharathv replied:
haha nice catch.. done.

dlex commented on src/v/serde/serde.h:
```diff
@@ -344,6 +408,28 @@ void write(iobuf& out, T t) {
+        //
+        // If we detect an overflow, we will clamp it to maximum supported
+        // duration, which is nanosecond::max() and if there is an underflow,
+        // we clamp it to minimum supported duration which is nanosecond::min().
```
Should we also mention that the clamping is temporary and currently there is intention to remove it in the future, throwing an exception in case of overflow?

bharathv replied:
Oh, I didn't know it was meant to be a temporary thing.. @dotnwat thats what you suggested?

dotnwat replied:
it's hard for me to imagine ever wanting to throw on overflow compared to clamping for duration types. but i'm sure there is some case out there where it makes sense. we can always revisit later.

dlex commented on src/v/serde/test/serde_test.cc:
```diff
@@ -829,3 +829,45 @@ SEASTAR_THREAD_TEST_CASE(type_str) {
+    constexpr milliseconds ms_neg{-1};
+    constexpr duration<int, std::kilo> ks(3);
+    constexpr duration<int16_t> int16t_d(10);
+    constexpr duration<int16_t> int32t_d(10);
```
Also `duration<int64_t,std::pico>` to check paths with `ratio::den!=1`?

bharathv replied:
Good idea.. done.. we need to be careful there as there is loss of precision converting to nanoseconds :laughing:

dotnwat replied:
nice idea.

bharathv commented on src/v/serde/serde.h:
```diff
@@ -166,9 +177,64 @@ inline constexpr auto const are_bytes_and_string_different = !(
+    using nano_period = std::chrono::nanoseconds::period;
+    using nano_rep = typename std::chrono::nanoseconds::rep;
+    // This is how duration_cast determines the output type.
+    using output_type = typename std::common_type<R, nano_rep, intmax_t>::type;
```
>It will probably be easier to just say that output_type is intmax_t. And don't care of what R or nano_rep are. Well only ensure that R is a signed integer.

Right.. this was just an additional conservative  check, retained because there is no runtime overhead.

> In theory R can even be double and the code below may correctly convert it into integer nanoseconds, but that would require additional checks

Agree.. check the static assert in L428 (I should probably move it here?), removed support for floating point types for this reason (and perf).

> unless we build the code with libc++ or with -std=gnu++17.

Btw we do build with libc++.

bharathv commented on src/v/serde/serde.h:
```diff
@@ -166,9 +177,64 @@ inline constexpr auto const are_bytes_and_string_different = !(
+        return duration.count() < 0 ? std::chrono::nanoseconds::min().count()
+                                    : std::chrono::nanoseconds::max().count();
+    }
+    // No overflow/underflow detected, we are safe to cast.
```
> (avoiding repeated multiplication in duration_cast)

I thought of doing this but skipped it because it would be consistent  to rely on the output of duration_cast<> logic in both directions and just rely minimally on the builtin for overflow checks. Also because the logic is already complex here and it would only confuse the reviewers. But I agree that its a good idea.

bharathv reviewed (commented):
will wait to hear back on comments before I push next rev ..
